src/analyses/pthreadAnalysis.ml

```diff
@@ -0,0 +1,1359 @@
+      |> List.group (compareBy (fst % fst))
+      |> flat_map fun_map
+    in
+    let globals = List.map Variable.show_def @@ Variables.get_globals () in
+    let promela =
+      let empty_line = "" in
+      let defs =
+        [ define "thread_count " ^ string_of_int thread_count
+        ; define "mutex_count " ^ string_of_int mutex_count
+        ; define "cond_var_count " ^ string_of_int cond_var_count
+        ; empty_line
+        ; define "checkStatus(op1, v, op2) " ^ checkStatus
+        ; empty_line
+        ; define "allTasks(prop) " ^ allTasks
+        ; empty_line
+        ; "#include \"../spin/pthread.base.pml\""
+        ; empty_line
+        ]
+      in
+      let init =
+        let run_threads =
+          (* NOTE: assumes no args are passed to the thread func *)
+          List.map
+            (fun t ->
+              let tid = Action.(t.tid) in
+              let thread_name = Option.get @@ Tbls.ThreadTidTbl.get_key tid in
+              let tid_str = string_of_int tid in
+              run thread_name ~arg:tid_str)
+            threads
+        in
+        let run_main = run promela_main ~arg:"0" in
+        let init_body = (run_main :: run_threads) @ [ "setReady(0);" ] in
+        [ "init {" ] @ List.map tabulate init_body @ [ "}" ]
+      in
+      let body =
+        let separator = [ empty_line ] in
+        List.flatten
+          [ defs
+          ; init
+          ; separator
+          ; fun_ret_defs
+          ; separator
+          ; globals
+          ; process_defs
+          ]
+      in
+      String.concat "\n" body
+    in
+    let dot_graph =
+      let dot_thread tid =
+        let show_edge (a, action, b) =
+          let show_node x =
+            "\"" ^ Resource.res_name tid ^ "_" ^ string_of_node x ^ "\""
+          in
+          show_node a
+          ^ "\t->\t"
+          ^ show_node b
+          ^ "\t[label=\""
+          ^ Action.to_pml action
+          ^ "\"]"
+        in
+        let subgraph_head =
+          "subgraph \"cluster_" ^ Resource.res_name tid ^ "\" {"
+        in
+        let subgraph_tail =
+          "label = \"" ^ Resource.res_name tid ^ "\";\n  }\n"
+        in
+        let edges_decls = Set.elements @@ Set.map show_edge @@ Edges.get tid in
+        (subgraph_head :: edges_decls) @ [ subgraph_tail ]
+      in
+      let lines =
+        Hashtbl.keys Edges.table
+        |> List.of_enum
+        |> List.unique
+        |> List.map dot_thread
+        |> List.concat
+      in
+      String.concat "\n  " ("digraph file {" :: lines) ^ "}"
+    in
+
+    Writer.write "promela model" "pml" promela ;
+    Writer.write "graph" "dot" dot_graph ;
+    print_endline
+      "Copy spin/pthread_base.pml to same folder and then do: spin -a \
+       pthread.pml && cc -o pan pan.c && ./pan"
+end
+
+module Spec : Analyses.MCPSpec = struct
+  module M = Messages
+  module List = BatList
+
+  (* Spec implementation *)
+  include Analyses.DefaultSpec
+
+  (** Domains *)
+  include PthreadDomain
```

Comment:
This is slightly unusual in our codebase. We're usually more explicit:
```suggestion
  module D = PthreadDomain.D
```
This also avoids including unnecessary stuff into this `Spec`, which could be just `open`ed for convenient access instead.